README
NAME
    Kvasir - Generic rule based processing engine

SYNOPSIS
    A simple example which just randomizes a number 10 times and outputs it
    to screen if it was under 500.

      use Kvasir::Constants;
      use Kvasir::Declare;
  
      my $i = 0;
  
      my $engine = engine {
          prehook "check_i" => does {
              $i++;
          
              return KV_ABORT if $i == 10;
              return KV_CONTINUE;
          };
      
          input "random" => does {
              return int(rand(1000));
          };
      
          rule "under_500" => does {
              my $input = $_[KV_INPUT];
              return KV_MATCH if $input->get("random") < 500;
              return KV_NO_MATCH;
          };
      
          action "mark_under_500" => does {
              my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
              $local->set("output" => $input->get("random") . " was under 500");
          };

          run "mark_under_500" => when "under_500";
      
          output "print" => does {
              my $local = $_[KV_LOCAL];
              my $output = $local->get("output");
              print $output, "\n" if defined $output;
          };
      }

    A fictional spam filter

      use Kvasir::Constants;
      use Kvasir::Declare;
  
      my $engine = engine {
          prehook "has_more_email" => instanceof "MyApp::HasMoreEmails";
      
          # Returns an Email::Simple instance
          input "email" => instanceof "MyApp::GetEmail";
      
          input "from" => does {
                my $input = $_[KV_INPUT];
                return $input->get("email")->header("From");
          };
      
          rule "is_spam" => instanceof "MyApp::SpamDetector";
      
          rule "is_from_boss" => does {
              my $input = $_[KV_INPUT];
          
              if ($input->get("from") =~ /theboss@company\.com/) {
                  return KV_MATCH;
              }
          
              return KV_NO_MATCH;
          };
      
          action "mark_for_deletion" => does {
              my $local = $_[KV_LOCAL];
              $local->set("delete-email" => 1);
          };
      
          run "mark_for_deletion" => when qw(is_spam is_from_boss);
      
          post_hook "process_mail" => does {
              my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
          
              if ($local->get("delete-mail") == 1) {
                  MyApp::EmailHandler->delete($input->get("email"));
              }
          
              return KV_CONTINUE;
          };
      };
  
      $engine->run();

DESCRIPTION
    Kvasir is a generic rule based processing engine. Processing is done
    while neither any pre nor post iteration hook aborts processing. Each
    engine supports multiple pre- and posthooks, inputs, output, actions and
    rules. An action is attached to one or several rules and is executed if
    any of the rules matches. Input is processed when needed and only once
    per iteration.

    Multiple engines can be run in parallel in a runloop. Runloops can be
    ran until no more data is available for processing or by a single step
    at a time. This way multiple runloops may be executed simultaneously.

USAGE
    Kvasir has a declarative interface via Kvasir::Declare which is the
    simplest way to define engines. The synopsis shows an exmaple that could
    be implemented (with reservations for syntax errors).

  Declarative interface
    Kvasir::Declare

  Loading engines from other sources
    Currently creation of engines are only possible using either the
    declarative interface or by creating "Kvasir::Engine"-objects directly.
    Future releases will allow loading of engine declarations via XML and
    other formats.

  Runloops
    Runloops transforms engines into something executable and runs them.

    Kvasir::Runloop

  Inputs, Outputs, Rules, Hooks, Actions
    The components that makes up an engine.

    Kvasir::Input, Kvasir::Output, Kvasir::Rule, Kvasir::Hook,
    Kvasir::Action

BUGS AND LIMITATIONS
    Please report any bugs or feature requests to "bug-kvasir@rt.cpan.org",
    or through the web interface at <http://rt.cpan.org>.

AUTHOR
    Claes Jakobsson "<claesjac@cpan.org>"

LICENCE AND COPYRIGHT
    Copyright (c) 2007, Versed Solutions "<info@versed.se>". All rights
    reserved.

    This software is released under the MIT license cited below.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 0.07 - 2007-10-08]

* with_args argument is now passed as a hash instead of a hash reference to entities
* Added 'defaults' and 'with_defaults' to declarative interface
* Added default argument sets to entities
* Corrected typo (post_hook => posthook) in Kvasir synopsis [DOC]
* Corrected POD names in InputUnderThreshold to actually match that module [DOC]

[Changes for 0.06 - 2007-10-02]

* Don't install Test::Kvasir::*
* Doesn't require the non-existing Perl 5.6.20 (5.00602) anymore. Spotted by David Cantrell [BUG]
* Added generic rules InputMatchesRegexp, InputOverThreshold and InputUnderThreshold

[Changes for 0.05 - 2007-10-01]

* Don't try to require a package if it's already defined [BUG]
* is_existing_package in Kvasir::Util should now work correctly [BUG]
* Corrected calling posthooks so that their argument positions are correct [BUG]
* Specifying a default global is now possible either via add_engine for Runloop or via run for Engine

[Changes for 0.04 - 2007-09-16]

* add_* in Engine now also accepts objects that'll be used to the specific item. Declarative interface
  is supports this using 'input => $object' etc.
* Added POD to Kvasir/Engine/* to make doc tests not complain that did does not contain any POD [DOC]

[Changes for 0.03 - 2007-09-16]

* Added some synopsises and better documentation [DOC]
* Removed some double docs from Engine.pm and cleaned it up a bit [DOC]
* Kvasir/Engine/*.pm doesn't set package anymore to avoid confusion and making them mixins [BUG/DOC]
 
[Changes for 0.02 - 2007-09-16]

* Inputs are now reset before each iteration [BUG]
* Added another example in synopsis so that peopledon't think this is a spam filter. [DOC]
* Updated documentation with some fixes. [DOC]
  
[Changes for 0.01 - 2007-09-15]

* Initial release to CPAN


lib/Kvasir.pm
package Kvasir;

use strict;
use warnings;

our $VERSION = "0.07";

1;
__END__

=head1 NAME

Kvasir - Generic rule based processing engine

=head1 SYNOPSIS

A simple example which just randomizes a number 10 times and outputs it to screen if it was under 500.

  use Kvasir::Constants;
  use Kvasir::Declare;
  
  my $i = 0;
  
  my $engine = engine {
      prehook "check_i" => does {
          $i++;
          
          return KV_ABORT if $i == 10;
          return KV_CONTINUE;
      };
      
      input "random" => does {
          return int(rand(1000));
      };
      
      rule "under_500" => does {
          my $input = $_[KV_INPUT];
          return KV_MATCH if $input->get("random") < 500;
          return KV_NO_MATCH;
      };
      
      action "mark_under_500" => does {
          my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
          $local->set("output" => $input->get("random") . " was under 500");
      };

      run "mark_under_500" => when "under_500";
      
      output "print" => does {
          my $local = $_[KV_LOCAL];
          my $output = $local->get("output");
          print $output, "\n" if defined $output;
      };
  }

A fictional spam filter

  use Kvasir::Constants;
  use Kvasir::Declare;
  
  my $engine = engine {
      prehook "has_more_email" => instanceof "MyApp::HasMoreEmails";
      
      # Returns an Email::Simple instance
      input "email" => instanceof "MyApp::GetEmail";
      
      input "from" => does {
            my $input = $_[KV_INPUT];
            return $input->get("email")->header("From");
      };
      
      rule "is_spam" => instanceof "MyApp::SpamDetector";
      
      rule "is_from_boss" => does {
          my $input = $_[KV_INPUT];
          
          if ($input->get("from") =~ /theboss@company\.com/) {
              return KV_MATCH;
          }
          
          return KV_NO_MATCH;
      };
      
      action "mark_for_deletion" => does {
          my $local = $_[KV_LOCAL];
          $local->set("delete-email" => 1);
      };
      
      run "mark_for_deletion" => when qw(is_spam is_from_boss);
      
      posthook "process_mail" => does {
          my ($input, $local) = @_[KV_INPUT, KV_LOCAL];
          
          if ($local->get("delete-mail") == 1) {
              MyApp::EmailHandler->delete($input->get("email"));
          }
          
          return KV_CONTINUE;
      };
  };
  
  $engine->run();

=head1 DESCRIPTION

Kvasir is a generic rule based processing engine. Processing is done while neither any pre nor post iteration hook 
aborts processing. Each engine supports multiple pre- and posthooks, inputs, output, actions and rules. An action is 
attached to one or several rules and is executed if any of the rules matches. Input is processed when needed and only 
once per iteration.

Multiple engines can be run in parallel in a runloop. Runloops can be ran until no more data is available for processing 
or by a single step at a time. This way multiple runloops may be executed simultaneously.

=head1 USAGE

Kvasir has a declarative interface via L<Kvasir::Declare> which is the simplest way to define engines. The synopsis shows 
an exmaple that could be implemented (with reservations for syntax errors).

=head2 Declarative interface

L<Kvasir::Declare>

=head2 Loading engines from other sources

Currently creation of engines are only possible using either the declarative interface or by 
creating C<Kvasir::Engine>-objects directly. Future releases will allow loading of engine declarations 
via XML and other formats.

=head2 Runloops

Runloops transforms engines into something executable and runs them.

L<Kvasir::Runloop>

=head2 Inputs, Outputs, Rules, Hooks, Actions

The components that makes up an engine.

L<Kvasir::Input>, L<Kvasir::Output>, L<Kvasir::Rule>, L<Kvasir::Hook>, L<Kvasir::Action>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to C<bug-kvasir@rt.cpan.org>, 
or through the web interface at L<http://rt.cpan.org>.

=head1 AUTHOR

Claes Jakobsson C<< <claesjac@cpan.org> >>

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2007, Versed Solutions C<< <info@versed.se> >>. All rights reserved.

This software is released under the MIT license cited below.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
